src/core/utils/slaac_address.cpp

```diff
@@ -262,47 +272,83 @@ void Slaac::GenerateIid(Ip6::NetifUnicastAddress &aAddress) const
      *  - RID is random (but stable) Identifier.
      *  - For pseudo-random function `F()` SHA-256 is used in this method.
      *  - `Net_Iface` is set to constant string "wpan".
-     *  - `Network_ID` is not used (optional per RF-7217).
+     *  - `Network_ID` is not used if `aNetworkId` is NULL (optional per RF-7217).
      *  - The `secret_key` is randomly generated on first use (using true
      *    random number generator) and saved in non-volatile settings for
      *    future use.
      *
      */
 
-    const uint8_t  netIface[] = {'w', 'p', 'a', 'n'};
-    uint16_t       dadCounter;
+    otError        error         = OT_ERROR_FAILED;
+    const uint8_t  netIface[]    = {'w', 'p', 'a', 'n'};
+    uint16_t       dadCounter    = 0;
+    uint16_t       endDadCounter = kMaxIidCreationAttempts;
     IidSecretKey   secretKey;
     Crypto::Sha256 sha256;
     uint8_t        hash[Crypto::Sha256::kHashSize];
 
     OT_STATIC_ASSERT(sizeof(hash) >= Ip6::Address::kInterfaceIdentifierSize,
                      "SHA-256 hash size is too small to use as IPv6 address IID");
 
+    if (aDadCounter)
+    {
+        dadCounter = *aDadCounter;
+
+        if ((UINT16_MAX - kMaxIidCreationAttempts) >= dadCounter)
+        {
+            endDadCounter = UINT16_MAX;
+        }
+        else
+        {
+            endDadCounter = dadCounter + kMaxIidCreationAttempts;
+        }
+    }
+
     GetIidSecretKey(secretKey);
 
-    for (dadCounter = 0; dadCounter < kMaxIidCreationAttempts; dadCounter++)
+    for (; dadCounter < endDadCounter; dadCounter++)
     {
         sha256.Start();
         sha256.Update(aAddress.mAddress.mFields.m8, BitVectorBytes(aAddress.mPrefixLength));
+
+        if (aNetworkId)
+        {
+            sha256.Update(aNetworkId, aNetworkIdLength);
+        }
+
         sha256.Update(netIface, sizeof(netIface));
         sha256.Update(reinterpret_cast<uint8_t *>(&dadCounter), sizeof(dadCounter));
         sha256.Update(secretKey.m8, sizeof(IidSecretKey));
         sha256.Finish(hash);
 
         aAddress.GetAddress().SetIid(&hash[0]);
 
-        // Exit and return the address if the IID is not reserved,
-        // otherwise, try again with a new dadCounter
+        // If the IID is reserved, try again with a new dadCounter
+        if (aAddress.GetAddress().IsIidReserved())
+        {
+            continue;
+        }
+
+        if (aDadCounter)
+        {
+            *aDadCounter = dadCounter;
+        }
 
-        VerifyOrExit(aAddress.GetAddress().IsIidReserved(), OT_NOOP);
+        // Exit and return the address if the IID is not reserved,
+        ExitNow(error = OT_ERROR_NONE);
     }
 
     otLogWarnUtil("SLAAC: Failed to generate a non-reserved IID after %d attempts", dadCounter);
```

Comment:
Strictly speaking, `dadCounter` no longer equals to the number of attempts. The number of attempts should always equal to `kMaxIidCreationAttempts`?

Reply:
good catch. will update.